### ISSUE TITLE
cannot call Lookup on a closed database

### ISSUE THREAD

**Issue by lutoma** [porech/caddy-maxmind-geolocation#17]
Hi,

thanks for the module, it works great generally speaking. However, after reloading Caddy, I start seeing lots of

```
caddy[3183269]: {"level":"warn","ts":1692868949.610951,"logger":"http.matchers.maxmind_geolocation","msg":"cannot lookup IP address","address":"<snip>:56631","error":"cannot call Lookup on a closed database"}
```

messages in the log and the matching stops working. Restarting Caddy fixes this (until the next reload).

This error seems to come directly from maxminddb-golang: https://github.com/oschwald/maxminddb-golang/blob/main/reader.go#L135

Just spitballing here, but I'm assuming Caddy closes the module/open file on reload somehow, but the module attempts to keep accessing the existing handle, or something like it?

We're using Caddy `v2.7.4 h1:J8nisjdOxnYHXlorUKXY75Gr6iBfudfoGhrJ8t7/flI=`, maybe this is a new interaction between newer versions of Caddy and the module? Let me know if there is any more useful info I could provide to help debug this.

**Comment by ghost** [comment]
I have the same problem on `v2.7.6 h1:w0NymbG2m9PcvKWsrXO6EEkY9Ru4FJK8uQbYcev1p3A=`. Not being able to reload one's config is a little annoying...but full restarts of Caddy don't take *too* long...

**Comment by ale-rinaldi** [comment]
Hello @lutoma and @wkeiuluf ,

that sucks, and what sucks the most is that I didn't find a way to reproduce this. I tried reloading Caddy several times and I still got successful results after the reload.

I took a look to maxminddb-golang and yes, it triggers the error if the buffer is nil. But the buffer is set to nil only if one explicitly calls Close(), and i call Close() only on cleanup.

So there's only a reason why this can happen: the module instance is used by Caddy after Cleanup() is called!

According to [the Caddy module lifecycle documentation](https://caddyserver.com/docs/extending-caddy#module-lifecycle), the cleanup is performed only when the module instance is no longer needed, and I'd expect this not to happen actually. It may be a Caddy bug that happens in some specific occasions, bug that's to be verified.

Meanwhile, I pushed a change to a branch that just tries to re-open the database if it has been closed; it's not my preferred way of doing this so I'm not sure I'll merge it, I'd prefer to better understand what's going on here that closes the file. However, I'd need your help to understand if this fixes things for you, since as I said I didn't find a way to reproduce.

Can some of you please try building Caddy with the version from the branch and confirm me that's fixed? Here's how to do it:

- Install [golang](https://go.dev/doc/install) and [xcaddy](https://github.com/caddyserver/xcaddy?tab=readme-ov-file#install)
- Clone this repository: `git clone https://github.com/porech/caddy-maxmind-geolocation.git`
- Go in the repository folder: `cd caddy-maxmind-geolocation` (you'll be on the main branch now)
- Build Caddy: `xcaddy build`
- You'll now find a `caddy` binary in the folder. Try running it in your environment and reloading; the problem should still be present
- Switch to the branch with the fix: `git checkout fix/lookup-on-closed-database`
- Build Caddy again: `xcaddy build`
- Run caddy again from the generated binary. The problem should be gone.

After ensuring this fixes things for you, I'll better understand what to do next.

Thank you!

**Comment by Daniel-Nldr** [comment]
I stumbled across this issue, while testing an upgrade from 2.6.4 to 2.7.x
After randomly clicking around the site eventually it would just be a blank white page, nothing showing in the dev console etc.

I went down the rabbit hole of building 2.7.3-2.7.6 with the required add-ons

The moment I went to 2.7.3, I had this issue, where it went blank, if I wanted long enough and tried a refresh I would get a page back and then enough refreshes/clicks and I'd once again get the blank page.
Interestingly enough, of the 3 browsers I had open and was refreshing on, 1 would die first and then a while later the other and often 1 would continue to work for some time still, which is a little odd

I then turned on and checked the error logs and sure enough I had this same message.

I've tested building 2.7.6 with your _lookup-on-closed-database_ branch, and I so far have not been able to reproduce the white page error

Hope this helps
